Bricks/sync/scope_owned.h

```diff
@@ -0,0 +1,388 @@
+    {
+      std::unique_lock<std::mutex> lock(mutex_);
+      condition_variable_.wait(lock, [this]() { return followers_.empty(); });
+    }
+  }
+
+  // Register ("increase the ref-count") a newly spawned follower, along with its termination signal callback.
+  // Must be called from within a mutex-locked section.
+  size_t RegisterFollowerFromLockedSection(std::function<void()> destruction_callback) {
+    followers_[++total_followers_spawned_throughout_lifetime_] = destruction_callback;
+    return total_followers_spawned_throughout_lifetime_;
+  }
+
+  // Unregister ("decrease the ref-count") a previously spawned follower.
+  // Locks its own mutex; must be called from outside a mutex-locked section.
+  void UnRegisterFollower(size_t follower_index) {
+    bool notify = false;
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      auto it = followers_.find(follower_index);
+      if (it == followers_.end()) {
+        CURRENT_THROW(AttemptedToUnregisterScopeOwnedBySomeoneElseMoreThanOnce());
+      }
+      followers_.erase(it);
+      notify = followers_.empty();
+    }
+    if (notify) {
+      condition_variable_.notify_one();
+    }
+  }
+
+  ActualInstance() = delete;
+  ActualInstance(const ActualInstance&) = delete;
+  ActualInstance(ActualInstance&&) = delete;
+  ActualInstance& operator=(const ActualInstance&) = delete;
+  ActualInstance& operator=(ActualInstance&&) = delete;
+
+  T instance_;                                          // The actual instance.
+  std::atomic_bool destructing_;                        // Whether it already is in the "destructing" mode.
+  std::mutex mutex_;                                    // The mutex to guard per-follower info.
+  std::condition_variable condition_variable_;          // The variable to notify as all the users are done.
+  std::map<size_t, std::function<void()>> followers_;   // Map of live follower -> termination signal callback.
+  size_t total_followers_spawned_throughout_lifetime_;  // A pre-increment of this value is the key in the map.
+};
+
+}  // namespace current::sync::impl
+
+template <typename T>
+class ScopeOwnedByMe;
+template <typename T>
+class ScopeOwnedBySomeoneElse;
+
+// `ScopeOwned<T>` is the  base class for `ScopeOwnedByMe<T>` and `ScopeOwnedBySomeoneElse<T>`.
+//
+// Unifies and simplifies:
+// 1) Basic usage patterns, such as `operator*()`, `operator->()`, and exceptions.
+// 2) Parameter passing and initialization, as the callee initializing its own `ScopeOwnedBySomeoneElse<T>`
+//    from the passed in parameter can accept `ScopeOwned<T>`, which would enable calling it with
+//    either `ScopeOwnedByMe<T>` or `ScopeOwnedBySomeoneElse<T>` as a parameter.
+template <typename T>
+class ScopeOwned {
+ private:
+  using T_INSTANCE = impl::ActualInstance<T>;
+  // `ScopeOwned<T>` is meant to be initialized by `ScopeOwnedByMe<T>` or `ScopeOwnedBySomeoneElse<T>` only.
+  // Its constructors are private to prevent accidental construction.
```

Comment:
I suggest making a note here about why you wouldn't just make the constructors protected.